nautobot/project-static/js/homepage_layout.js

```diff
@@ -0,0 +1,130 @@
+
+        // Toggle the collapsed class
+        if (collapsibleDiv.classList.contains('collapsed')) {
+            collapsibleDiv.classList.remove('collapsed');
+        } else {
+            collapsibleDiv.classList.add('collapsed');
+        }
+
+        // Update the state in localStorage
+        var isCollapsed = collapsibleDiv.classList.contains('in');
+        localStorage.setItem(elementId, isCollapsed ? 'collapsed' : 'expanded');
+        // Set Cookie value based on isCollapsed
+        if (isCollapsed) {
+            document.cookie = elementId + "=True; path=/";
+        } else {
+            document.cookie = elementId + "=False; path=/";
+        }
+    }
+
+    // Add event listener to each collapsible div
+    var collapseIcons = document.querySelectorAll('.collapse-icon');
+    collapseIcons.forEach(function(icon) {
+        icon.addEventListener('click', function() {
+            var elementId = this.id;
+            toggleAndSaveState(elementId);
+        });
+    });
+});
+
+// Enables panels on homepage to be rearranged via drag and drop
+document.addEventListener('DOMContentLoaded', function() {
+    // Get the container element
+    var draggableHomepagePanels = document.getElementById('draggable-homepage-panels');
+
+    // Add event listeners for drag and drop events
+    draggableHomepagePanels.addEventListener('dragstart', handleDragStart, false);
+    draggableHomepagePanels.addEventListener('dragover', handleDragOver, false);
+    draggableHomepagePanels.addEventListener('drop', handleDrop, false);
+
+    // Load saved panel order on page load
+    loadSavedPanelOrder();
+
+    // Function to handle the start of a drag event
+    function handleDragStart(e) {
+        e.dataTransfer.setData('text/plain', e.target.id);
+        e.target.classList.add('dragging');
+    }
+
+    // Function to handle a drag over event
+    function handleDragOver(e) {
+        e.preventDefault();
+    }
+
+    // Function to handle a drop event
+    function handleDrop(e) {
+        e.preventDefault();
+        var draggedId = e.dataTransfer.getData('text/plain');
+        var droppedOn = e.target.closest('.panel'); // Get the closest panel to the drop location
+
+        if (droppedOn) { // Check if a panel was found
+            var droppedOnId = droppedOn.id;
+            swapElements(document.getElementById(draggedId), document.getElementById(droppedOnId));
+            savePanelOrder();
+        }
+        
+        // Remove the "dragging" class from the panel that was dragged
+        document.getElementById(draggedId).classList.remove('dragging');
+    }
+
+    // Function to swap two elements
+    function swapElements(obj1, obj2) {
+        var temp = document.createElement("div");
+        obj2.parentNode.insertBefore(temp, obj2);
+        obj1.parentNode.insertBefore(obj2, obj1);
+        temp.parentNode.insertBefore(obj1, temp);
+        temp.parentNode.removeChild(temp);
+    }
```

Comment:
It "feels" a bit weird to me that drag-and-drop works as a "swap" operation. I'd have expected it to be more of a "move" or "insert" operation, i.e the dragged element gets inserted before/after the element it's dropped onto, not that the two elements exchange positions. Maybe just me?

Reply:
I found other ways using jQuery, but this was a common way I found online when looking this up.  Functionally it does what I need, so I'm a little confused.  Is this way a concern, or do you know of a better way?  I'm not as experienced with JS and didn't find many other ways to do so online, so guide me Obi Wan Kenobi.

Reply:
My concern wasn't with the code but with the behavior - it feels off to me that the element you drop onto also gets relocated, I would have expected just the dragged element would be relocated (other elements would shift to make room for it, but the order of the other elements wouldn't change).

Maybe replace this function with something as simple as like `draggedElement.parentNode.insertBefore(targetElement)` perhaps - not tested, I'll see if I can make time to try it out.

Reply:
I actually tried that this morning and it worked, but I think the behavior wasn't exactly the same.  I will test it again later today and let you know the results.

Reply:
I was hoping for different behavior, in case that wasn't clear. :smile: Just my opinion, of course!

Reply:
Marking as resolved after implementing Glenn's `handleDrop()` function rewrite